[WGSL] Add pass to rewrite entry points to be compatible with Metal

tadeuzagallo opened the pull request:
#### 021a45c18eba49cfc45b5cad694f05095c138414
<pre>
[WGSL] Add pass to rewrite entry points to be compatible with Metal
<a href="https://bugs.webkit.org/show_bug.cgi?id=250832">https://bugs.webkit.org/show_bug.cgi?id=250832</a>
&lt;rdar://problem/104425157&gt;

Reviewed by Myles C. Maxfield.

For entry point function, builtin inputs contained in structs must be hoisted
into function parameters, and any non-builtin parameters must be moved into a
`stage_in` struct.

The implementation can&apos;t quite handle absolutely every use case yet, but the
patch was getting pretty large so I thought I pause here and commit something
that works.

A couple other things had to happen to get this initial implementation working:
- I needed some basic type information, so I added a very rudimentary pass to
  resolve type names and a new struct type.
- I also had to modify the AST to use `Ref`/`RefPtr` instead of `UniqueRef`/`unique_ptr`
  for types, since we need to be able to copy nodes when moving parameters around.
- Fixed a couple small serialization bugs with missing semicolons in MetalFunctionWriter
  and added serialization for matrix types.

* Source/WebGPU/WGSL/AST/ASTAttribute.h:
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h:
* Source/WebGPU/WGSL/AST/ASTCallableExpression.h:
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::ParameterizedType::ParameterizedType):
(isType):
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp: Added.
(WGSL::EntryPointRewriter::EntryPointRewriter):
(WGSL::EntryPointRewriter::getResolvedType):
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::collectParameters):
(WGSL::EntryPointRewriter::constructInputStruct):
(WGSL::EntryPointRewriter::materialize):
(WGSL::EntryPointRewriter::visit):
(WGSL::EntryPointRewriter::appendBuiltins):
(WGSL::RewriteEntryPoints::RewriteEntryPoints):
(WGSL::RewriteEntryPoints::visit):
(WGSL::rewriteEntryPoints):
* Source/WebGPU/WGSL/EntryPointRewriter.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseAttribute):
(WGSL::Parser&lt;Lexer&gt;::parseTypeDecl):
(WGSL::Parser&lt;Lexer&gt;::parseTypeDeclAfterIdentifier):
(WGSL::Parser&lt;Lexer&gt;::parseArrayType):
(WGSL::Parser&lt;Lexer&gt;::parseVariableDeclWithAttributes):
(WGSL::Parser&lt;Lexer&gt;::parseFunctionDecl):
(WGSL::Parser&lt;Lexer&gt;::parseStatement):
(WGSL::Parser&lt;Lexer&gt;::parseUnaryExpression):
(WGSL::Parser&lt;Lexer&gt;::parsePrimaryExpression):
(WGSL::Parser&lt;Lexer&gt;::parseCoreLHSExpression):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/PhaseTimer.h: Added.
(WGSL::dumpASTIfNeeded):
(WGSL::dumpASTAfterParsingIfNeeded):
(WGSL::dumpASTBetweenEachPassIfNeeded):
(WGSL::dumpASTAtEndIfNeeded):
(WGSL::logPhaseTimes):
(WGSL::PhaseTimer::PhaseTimer):
(WGSL::PhaseTimer::~PhaseTimer):
* Source/WebGPU/WGSL/ResolveTypeReferences.cpp: Added.
(WGSL::ResolveTypeReferences::ResolveTypeReferences):
(WGSL::ResolveTypeReferences::visit):
(WGSL::resolveTypeReferences):
* Source/WebGPU/WGSL/ResolveTypeReferences.h: Added.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: <a href="https://commits.webkit.org/259149@main">https://commits.webkit.org/259149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/978fcfeff1b3d74ab0ad5cb46b1272d34bd91d82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104058 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13153 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36987 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113270 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173577 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108002 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4065 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96294 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112351 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109828 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/10954 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94015 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38626 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92801 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25627 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80283 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6522 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27004 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6667 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3544 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12664 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46538 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6301 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8444 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -28,24 +28,33 @@ async function helloTriangle() {
-                         return vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+                         var vout : Vertex;
+                         vout.Position = vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+                         vout.color = vec4<f32>(pos[VertexIndex], 0.0, 1.0);
```
We're still missing the addition here, but the parser doesn't support it yet

litherum commented on Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
```diff
@@ -27,12 +27,18 @@
+
+class StructDecl;
+
+class TypeDecl : public Node, public ThreadSafeRefCounted<TypeDecl> {
```
Threadsafe? Really?

tadeuzagallo replied:
I guess it was force of habit. I'll remove the ThreadSafe.

litherum commented on Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
```diff
@@ -143,7 +157,42 @@ class ParameterizedType : public TypeDecl {
 };
 
-} // namespace WGSL::AST
+class StructType final : public TypeDecl {
```
Does it really make sense to have all these classes jammed together in the same file?

djg replied:
I have a refactoring where I split them out. I should get to landing all that stuff.

litherum commented on Source/WebGPU/WGSL/EntryPointRewriter.cpp:
```diff
@@ -0,0 +1,284 @@
+    String m_structParameterName;
+};
+
+AST::TypeDecl& EntryPointRewriter::getResolvedType(AST::TypeDecl& type)
```
I'm kind of uninformed, but does this work for types that are trees of other types? Like array<f32, 5>?

tadeuzagallo replied:
This function should recursively resolve those eventually, e.g.

```
type v = vec;
type f = f32;
v<f, 5>
```

should resolve to `vec<f32, 5>`

litherum commented on Source/WebGPU/WGSL/EntryPointRewriter.cpp:
```diff
@@ -0,0 +1,284 @@
+
+void EntryPointRewriter::rewrite()
+{
+    m_structTypeName = makeString("__", m_functionDecl.name(), "_inT");
+    m_structParameterName = makeString("__", m_functionDecl.name(), "_in");
```
We should make sure that m_functionDecl.name() is the mangled name, rather than the name in the original source. Also, how are we going to guarantee there won't be any name collisions?

tadeuzagallo replied:
That's the nice thing about doing this as an AST pass. As long as the mangling pass runs after this, we don't have to worry about that. As for the name collisions, the spec says that there can't be two functions with the same name, and the `__` prefix is invalid for user defined values, so I think this should work?

litherum commented on Source/WebGPU/WGSL/WGSL.cpp:
```diff
@@ -58,8 +80,25 @@ SuccessfulCheck::~SuccessfulCheck() = default;
+        PhaseTimer phaseTimer("prepare total", phaseTimes);
+
+        RUN_PASS(resolveTypeReferences, ast);
+        RUN_PASS(rewriteEntryPoints, ast);
```
Shouldn't this call be in a MSL-specific area? If we add a second backend code generator, surely that second backend wouldn't want this MSL-specific rewriting, right?

tadeuzagallo replied:
I guess Metal-specific, not MSL, right? But in any case, we can always conditionally remove the pass later. I'm just trying to keep things as simple as possible (within reason) for now.

litherum commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -28,24 +28,33 @@ async function helloTriangle() {
                              vec2<f32>( 0.5, -0.5)
                          );
-                         return vec4<f32>(pos[VertexIndex], 0.0, 1.0);
+                         var vout : Vertex;
```
`vout`? Usually we try to use full words

tadeuzagallo replied:
sure, I  can rename, but I just copied this from a couple lines below.

litherum commented on Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
```diff
@@ -28,24 +28,33 @@ async function helloTriangle() {
                      }
     `;
- */
+/*
```
Ditto about leaving code commented; usually we either user it or delete it.

tadeuzagallo replied:
All of these examples files have both the MSL and WGSL code, one which is commented out. Should I just delete the MSL code once we switched for the WGSL for each file?

litherum commented on Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
```diff
@@ -74,9 +81,16 @@ class NamedType final : public TypeDecl {
+    TypeDecl* maybeResolvedReference() const { return m_resolvedReference.get(); }
 
 private:
+    void resolveTypeReference(Ref<TypeDecl>&& typeDecl)
+    {
+        m_resolvedReference = WTFMove(typeDecl);
+    }
```
I must not be understanding this. The `NamedType` is a superclass of a struct declaration, right? I'm confused about why a TypeDecl has a ref to another TypeDecl for the resolved reference... Why isn't this specific to typedefs? Also, how do type parameters work?

tadeuzagallo replied:
`NamedType` isn't a superclass of `StructDecl`. This is to cover the following:

```
struct S { }
fn f(s: S) { }
```

What happens now is:
1. We read the parameter `s` from `f`,  then get its type `S`;
2. We want to be able to access the actual struct so we can read its fields' types.

Before this patch, all we got was a `NamedType { name = 'S' }`, now we have a way to go from that to `StructType { .members = { ... } }` so we can read the actual fields of `S`.

This is unrelated to type parameters, that will come later.

litherum commented on Source/WebGPU/WGSL/PhaseTimer.h:
```diff
@@ -0,0 +1,110 @@
+    return false;
+}
+
+static inline bool dumpASTAfterParsingIfNeeded(AST::ShaderModule& program)
+{
+    return dumpASTIfNeeded(dumpASTAfterParsing, program, "AST after parsing");
+}
+
+static inline bool dumpASTBetweenEachPassIfNeeded(AST::ShaderModule& program, const char* message)
+{
+    return dumpASTIfNeeded(dumpASTBeforeEachPass, program, message);
+}
+
+static inline bool dumpASTAtEndIfNeeded(AST::ShaderModule& program)
+{
+    return dumpASTIfNeeded(dumpASTAtEnd, program, "AST at end");
+}
```
Do these really need to be separate functions?

tadeuzagallo replied:
Honestly, this was just borrowed verbatim from the WHLSL implementation. Not sure if it really needs to be as is, but it didn't bother me

litherum commented on Source/WebGPU/WGSL/PhaseTimer.h:
```diff
@@ -0,0 +1,110 @@
+static inline bool dumpASTIfNeeded(bool shouldDump, AST::ShaderModule& program, const char* message)
+{
+    if (UNLIKELY(shouldDump)) {
+        dataLogLn(message);
```
Ditto about WTFLogAlways

tadeuzagallo replied:
If you're not strongly opposed, I'd prefer to stick to `dataLog`. It allows us to implement custom print functions for our data types, which makes it a lot easier to dump data structures (and we are already using it right below to dump the AST anyway, so we'd be just changing the string before the AST to use a different log function)

litherum reviewed (approved):
Much of this is over my head, but ok.

tadeuzagallo committed:
[WGSL] Add pass to rewrite entry points to be compatible with Metal
https://bugs.webkit.org/show_bug.cgi?id=250832
<rdar://problem/104425157>

Reviewed by Myles C. Maxfield.

For entry point function, builtin inputs contained in structs must be hoisted
into function parameters, and any non-builtin parameters must be moved into a
`stage_in` struct.

The implementation can't quite handle absolutely every use case yet, but the
patch was getting pretty large so I thought I pause here and commit something
that works.

A couple other things had to happen to get this initial implementation working:
- I needed some basic type information, so I added a very rudimentary pass to
  resolve type names and a new struct type.
- I also had to modify the AST to use `Ref`/`RefPtr` instead of `UniqueRef`/`unique_ptr`
  for types, since we need to be able to copy nodes when moving parameters around.
- Fixed a couple small serialization bugs with missing semicolons in MetalFunctionWriter
  and added serialization for matrix types.

* Source/WebGPU/WGSL/AST/ASTAttribute.h:
* Source/WebGPU/WGSL/AST/ASTBuiltinAttribute.h:
* Source/WebGPU/WGSL/AST/ASTCallableExpression.h:
* Source/WebGPU/WGSL/AST/ASTForward.h:
* Source/WebGPU/WGSL/AST/ASTFunctionDecl.h:
* Source/WebGPU/WGSL/AST/ASTNode.h:
* Source/WebGPU/WGSL/AST/ASTStringDumper.cpp:
(WGSL::AST::StringDumper::visit):
* Source/WebGPU/WGSL/AST/ASTStringDumper.h:
* Source/WebGPU/WGSL/AST/ASTStructureDecl.h:
* Source/WebGPU/WGSL/AST/ASTTypeDecl.h:
(WGSL::AST::ParameterizedType::ParameterizedType):
(isType):
* Source/WebGPU/WGSL/AST/ASTVariableDecl.h:
* Source/WebGPU/WGSL/AST/ASTVisitor.cpp:
(WGSL::AST::Visitor::visit):
* Source/WebGPU/WGSL/AST/ASTVisitor.h:
* Source/WebGPU/WGSL/EntryPointRewriter.cpp: Added.
(WGSL::EntryPointRewriter::EntryPointRewriter):
(WGSL::EntryPointRewriter::getResolvedType):
(WGSL::EntryPointRewriter::rewrite):
(WGSL::EntryPointRewriter::collectParameters):
(WGSL::EntryPointRewriter::constructInputStruct):
(WGSL::EntryPointRewriter::materialize):
(WGSL::EntryPointRewriter::visit):
(WGSL::EntryPointRewriter::appendBuiltins):
(WGSL::RewriteEntryPoints::RewriteEntryPoints):
(WGSL::RewriteEntryPoints::visit):
(WGSL::rewriteEntryPoints):
* Source/WebGPU/WGSL/EntryPointRewriter.h: Added.
* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::visit):
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseAttribute):
(WGSL::Parser<Lexer>::parseTypeDecl):
(WGSL::Parser<Lexer>::parseTypeDeclAfterIdentifier):
(WGSL::Parser<Lexer>::parseArrayType):
(WGSL::Parser<Lexer>::parseVariableDeclWithAttributes):
(WGSL::Parser<Lexer>::parseFunctionDecl):
(WGSL::Parser<Lexer>::parseStatement):
(WGSL::Parser<Lexer>::parseUnaryExpression):
(WGSL::Parser<Lexer>::parsePrimaryExpression):
(WGSL::Parser<Lexer>::parseCoreLHSExpression):
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/PhaseTimer.h: Added.
(WGSL::dumpASTIfNeeded):
(WGSL::dumpASTAfterParsingIfNeeded):
(WGSL::dumpASTBetweenEachPassIfNeeded):
(WGSL::dumpASTAtEndIfNeeded):
(WGSL::logPhaseTimes):
(WGSL::PhaseTimer::PhaseTimer):
(WGSL::PhaseTimer::~PhaseTimer):
* Source/WebGPU/WGSL/ResolveTypeReferences.cpp: Added.
(WGSL::ResolveTypeReferences::ResolveTypeReferences):
(WGSL::ResolveTypeReferences::visit):
(WGSL::resolveTypeReferences):
* Source/WebGPU/WGSL/ResolveTypeReferences.h: Added.
* Source/WebGPU/WGSL/WGSL.cpp:
(WGSL::prepare):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WGSL/ParserTests.cpp:
(TestWGSLAPI::TEST):
* Websites/webkit.org/demos/webgpu/scripts/hello-triangle.js:
(async helloTriangle):

Canonical link: https://commits.webkit.org/259149@main
